src/compiler/checker.ts

```diff
@@ -6753,9 +6763,132 @@ namespace ts {
+            }
+            if (!left) {
+                return right!;
+            }
+            if (!right || left === right) {
+                return left;
+            }
+            const names = createMap<true>();
+            // We join names in an upperCamelCase fashion.
+            const seperator = "And";
+            const leftNames = unescapeLeadingUnderscores(left).split(seperator);
+            const rightNames = unescapeLeadingUnderscores(right).split(seperator);
+            let first = true;
+            for (const list of [leftNames, rightNames]) {
+                for (const n of list) {
+                    names.set(first ? (first = false, n) : (n.slice(0, 1).toUpperCase() + n.slice(1)), true);
+                }
+            }
+            if (names.size > 5) {
+                // Combining >5 names just looks bad
+                return `arg${index}` as __String;
+            }
+            return escapeLeadingUnderscores(arrayFrom(names.keys()).join(seperator));
+        }
+
+        function combineUnionParameters(left: Signature, right: Signature) {
+            const longest = getParameterCount(left) >= getParameterCount(right) ? left : right;
+            const shorter = longest === left ? right : left;
+            const longestCount = getParameterCount(longest);
+            const eitherHasEffectiveRest = (hasEffectiveRestParameter(left) || hasEffectiveRestParameter(right));
+            const needsExtraRestElement = eitherHasEffectiveRest && !hasEffectiveRestParameter(longest);
+            const params = new Array<Symbol>(longestCount + (needsExtraRestElement ? 1 : 0));
+            let i = 0;
+            for (; i < longestCount; i++) {
+                const longestParamType = tryGetTypeAtPosition(longest, i)!;
+                const shorterParamType = tryGetTypeAtPosition(shorter, i) || unknownType;
+                let unionParamType = getIntersectionType([longestParamType, shorterParamType]);
+                const isRestParam = eitherHasEffectiveRest && !needsExtraRestElement && i === (longestCount - 1);
+                if (isRestParam) {
+                    unionParamType = createArrayType(unionParamType);
+                }
+                const isOptional = i > getMinArgumentCount(longest);
+                const leftName = tryGetNameAtPosition(left, i);
+                const rightName = tryGetNameAtPosition(right, i);
+                const paramSymbol = createSymbol(SymbolFlags.FunctionScopedVariable | (isOptional && !isRestParam ? SymbolFlags.Optional : 0), combinePartameterNames(i, leftName, rightName));
+                paramSymbol.type = unionParamType;
+                params[i] = paramSymbol;
+            }
+            if (needsExtraRestElement) {
+                const restParamSymbol = createSymbol(SymbolFlags.FunctionScopedVariable, "args" as __String);
+                restParamSymbol.type = createArrayType(getTypeAtPosition(shorter, i));

```

Comment:
Can't you just use `longestCount` here instead of the funniness with `i` from the loop.